src/dtls.c

```diff
@@ -290,60 +343,385 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+
+                InitCipherSpecs(&nonConstSSL->specs);
+
+                nonConstSSL->session->sessionIDSz = 0;
+                nonConstSSL->options.cipherSuite0 = 0;
+                nonConstSSL->options.cipherSuite = 0;
+                nonConstSSL->extensions = sslExts;
+
+                nonConstSSL->options.tls = 0;
+                nonConstSSL->options.tls1_1 = 0;
+                nonConstSSL->options.tls1_3 = 0;
+            }
+dtls13_cleanup:
+            TLSX_FreeAll(parsedExts, ssl->heap);
+        }
+        else
+            ret = SendAlert((WOLFSSL*)ssl, alert_fatal, illegal_parameter);
+#else
+#error "WOLFSSL_SEND_HRR_COOKIE has to be defined to use DTLS 1.3 server"
+#endif
+    }
+    else
+#endif
+    {
+        if (!ch->dtls12cookieSet) {
+            ret = CreateDtls12Cookie(ssl, ch, ch->dtls12cookie);
+            if (ret != 0)
+                return ret;
+            ch->dtls12cookieSet = 1;
+        }
+        ret = SendHelloVerifyRequest((WOLFSSL*)ssl, ch->dtls12cookie,
+                DTLS_COOKIE_SZ);
+    }
+    return ret;
+}
+
+int DoClientHelloStateless(WOLFSSL* ssl, const byte* input,
+                           word32* inOutIdx, word32 helloSz)
 {
-    byte cookie[DTLS_COOKIE_SZ];
     int ret;
     WolfSSL_CH ch;
+    byte isTls13 = 0;
+    PskInfo pskInfo;
 
-    *process = 1;
+    XMEMSET(&pskInfo, 0, sizeof(pskInfo));
+    XMEMSET(&ch, 0, sizeof(ch));
+
+    ssl->options.dtlsStateful = 0;
     ret = ParseClientHello(input + *inOutIdx, helloSz, &ch);
     if (ret != 0)
         return ret;
 
 #ifdef WOLFSSL_DTLS_NO_HVR_ON_RESUME
-    {
-        byte isValid = 0;
-        ret = TlsResumptionIsValid(ssl, &ch, &isValid);
+    ret = TlsResumptionIsValid(ssl, &ch, &pskInfo);
+    if (ret != 0)
+        return ret;
+    if (pskInfo.isValid) {
+        ssl->options.dtlsStateful = 1;
+        return 0;
+    }
+#endif /* WOLFSSL_DTLS_NO_HVR_ON_RESUME */
+
+#ifdef WOLFSSL_DTLS13
+    if (IsAtLeastTLSv1_3(ssl->version)) {
+        ret = TlsCheckSupportedVersion(ssl, &ch, &isTls13, &pskInfo);
         if (ret != 0)
             return ret;
-        if (isValid)
-            return 0;
+        if (isTls13) {
+            ret = TlsxFindByType(&ch.cookie, TLSX_COOKIE, ch.extension);
+            if (ret != 0)
+                return ret;
+        }
     }
-#endif /* WOLFSSL_DTLS_NO_HVR_ON_RESUME */
+#endif
 
-    ret = CreateDtlsCookie(ssl, &ch, cookie);
-    if (ret != 0)
-        return ret;
-    if (ch.cookie.size != DTLS_COOKIE_SZ ||
-        XMEMCMP(ch.cookie.elements, cookie, DTLS_COOKIE_SZ) != 0) {
-        *process = 0;
-        ret = SendHelloVerifyRequest(ssl, cookie, DTLS_COOKIE_SZ);
+    if (ch.cookie.size == 0) {
+        ret = SendStatelessReply((WOLFSSL*)ssl, &ch, isTls13, &pskInfo);
+    }
+    else {
+        byte cookieGood;
+        ret = CheckDtlsCookie(ssl, &ch, isTls13, &cookieGood);
+        if (ret != 0)
+            return ret;
+        if (!cookieGood)
```

Comment:
The alert is checked and sent in `SendStatelessReply`. I'll refactor to move it closer to the check.

Reply:
```
https://www.rfc-editor.org/rfc/rfc6347#section-4.2.1
   If a server receives a ClientHello with an invalid cookie, it SHOULD
   treat it the same as a ClientHello with no cookie.  This avoids
   race/deadlock conditions if the client somehow gets a bad cookie
   (e.g., because the server changes its cookie signing key).

   Note to implementors: This may result in clients receiving multiple
   HelloVerifyRequest messages with different cookies.  Clients SHOULD
   handle this by sending a new ClientHello with a cookie in response to
   the new HelloVerifyRequest.
```
No alert for 1.2.

Reply:
no, in 1.2 we should keep sending cookies

Reply:
you didn't move the check closer in the end.

Reply:
Moving it closer in next push.